llvm/utils/TableGen/SubtargetEmitter.cpp

```
@@ -1781,13 +1781,68 @@ void SubtargetEmitter::EmitHwModeCheck(const std::string &ClassName,
+  OS << "  // Collect HwModes and store them in bits\n";
+  OS << "  unsigned Modes = 0;\n";
   for (unsigned M = 1, NumModes = CGH.getNumModeIds(); M != NumModes; ++M) {
     const HwMode &HM = CGH.getMode(M);
-    OS << "  if (checkFeatures(\"" << HM.Features << "\")) return " << M
-       << ";\n";
+    OS << "  if (checkFeatures(\"" << HM.Features << "\")) Modes |= (1 << "
+       << (M - 1) << ");\n";
   }
-  OS << "  return 0;\n}\n";
+  OS << "  return Modes;\n}\n";
+  // End emitting for getHwModeSet().
+
+  // Start emitting for getHwMode().
+  OS << "unsigned " << ClassName
+     << "::getHwMode(enum HwModeType type) const {\n";
+  OS << "  unsigned Modes = getHwModeSet();\n";
+  OS << "\n  if(!Modes)\n    return Modes;\n\n";
+  OS << "  switch (type) {\n";
+  OS << "  case HwMode_Default:\n    return llvm::countr_zero(Modes) + 1;\n"
+     << "    break;\n";
+  OS << "  case HwMode_ValueType: {\n    Modes &= " << ValueTypeModes << ";\n"
+     << "    if (!Modes)\n      return Modes;\n"
+     << "    if (!llvm::has_single_bit<unsigned>(Modes))\n"
+     << "      llvm_unreachable(\"Two or more HwModes for ValueType were "
+        "found!\");\n"
+     << "    return llvm::countr_zero(Modes) + 1;\n  }\n";
+  OS << "  case HwMode_RegInfo: {\n    Modes &= " << RegInfoModes << ";\n"
+     << "    if (!Modes)\n      return Modes;\n"
+     << "    if (!llvm::has_single_bit<unsigned>(Modes))\n"
+     << "      llvm_unreachable(\"Two or more HwModes for RegInfo were "
+        "found!\");\n"
+     << "    return llvm::countr_zero(Modes) + 1;\n  }\n";
+  OS << "  case HwMode_EncodingInfo: {\n    Modes &= " << EncodingInfoModes
+     << ";\n"
+     << "    if (!Modes)\n      return Modes;\n"
+     << "    if (!llvm::has_single_bit<unsigned>(Modes))\n"
+     << "      llvm_unreachable(\"Two or more HwModes for Encoding were "
+        "found!\");\n"
+     << "    return llvm::countr_zero(Modes) + 1;\n  }\n";
```

All three of the non-Default cases are essentially identical other than the mask value. We should collapse this into ~1/3 the code.